File: 21-merge-two-sorted-lists/merge-two-sorted-lists.ts
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */

function mergeTwoLists(list1: ListNode | null, list2: ListNode | null): ListNode | null {
    if(list1 === null) return list2;
    if(list2 === null) return list1;
    let list : ListNode = new ListNode();
    if(list1.val>list2.val){
        list = list2;
        list2 = list2.next;
    }else{
        list = list1;
        list1 = list1.next;
        }
    let head = list;
    let head1 = list1;
    let head2 = list2;
    while(head1 !== null && head2 !== null){
        if(head1.val <= head2.val){
            head.next = head1;
            head = head1;
            head1 = head1.next;
        }
        else{
            head.next = head2;
            head = head2;
            head2 = head2.next;
        }
    }
    if(head1 === null) head.next = head2;
    else if(head2 === null) head.next = head1;
    return list;
    
};

File: 21-merge-two-sorted-lists/merge-two-sorted-lists.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode mergeTwoLists(ListNode list1, ListNode list2) {
        ListNode dummyHead = new ListNode();
        ListNode cur = dummyHead;
        ListNode curNode1 = list1;
        ListNode curNode2 = list2;
        while(curNode1 != null && curNode2 != null) {
            if(curNode1.val < curNode2.val){
                cur.next = curNode1;
                cur = cur.next;
                curNode1 = curNode1.next;
            }
            else{
                cur.next = curNode2;
                cur = cur.next;
                curNode2 = curNode2.next;
            }
        }
        if(curNode1 != null) cur.next= curNode1;
        if(curNode2 != null) cur.next = curNode2;
        return dummyHead.next;
    }
}